### We change the signal by trading the deviations of the 30-day moving average of the markets spot rates to the Real Exchange Rates we computed

We assume the following theoretical value of the interest rate : $St = P(usd)/P(f) * GDP(f)/GDP(usd)$

In [11]:
import pandas as pd

#Importing PPP data 
from data import PPP_import

df_PPP = PPP_import().dropna() #We remove any date exhibiting Nan, we'll need to relax this hyptohesis

#Importing GDP data
df_GDP = pd.read_csv("/Users/hirecheariles/Documents/Cours/Master Finance/ARP/Data/GDPPPPData.csv", sep=";", skiprows=4)

df_GDP.drop(["Indicator Name", "Indicator Code", "2025", "Unnamed: 70"], axis=1, inplace = True)

df_GDP = df_GDP.loc[:, ["Country Code"]].join(df_GDP.loc[:, "1990":]).set_index("Country Code").T #Only keeping relevant columns
US_GDP = df_GDP.loc[:, "USA"] #Storing the reference GDP
df_GDP = df_GDP.loc[:, ["AUS", "BRA", "CAN", "CHL", "EMU", "HUN", "IND", "IDN", "ISR", "JPN", "MEX", "NZL", "NOR", "POL", "RUS", "SGP", "ZAF", "SWE", "CHE", "CZE", "GBR"]]

GDPmapping = pd.DataFrame(index = ["AUS", "BRA", "CAN", "CHL", "EMU", "HUN", "IND", "IDN", "ISR", "JPN", "MEX", "NZL", "NOR", "POL", "RUS", "SGP", "ZAF", "SWE", "CHE", "CZE", "GBR"],
                       data = ['AUD', 'BRL', 'CAD', 'CLP', 'EUR', 'HUF', "INR", "IDR", "ILS", "JPY", "MXN", "NZD", "NOK", "PLN", "RUB", "SGD", "ZAR", "SEK", "CHF", "CZK", "GBP"]).to_dict()[0]

df_GDP = df_GDP.rename(columns=GDPmapping) #Right names to the right columns
df_GDP = df_GDP.div(US_GDP, axis=0) #we compute the ratio GDP(f)/GDP(usd)
df_GDP.index = pd.to_datetime(df_GDP.index, format="%Y") #changing to datetime

df_GDP.head()

Country Code,AUD,BRL,CAD,CLP,EUR,HUF,INR,IDR,ILS,JPY,...,NZD,NOK,PLN,RUB,SGD,ZAR,SEK,CHF,CZK,GBP
1990-01-01,0.049753,0.167271,0.093924,0.010015,0.901209,0.015951,0.175802,0.093670,0.012286,0.412311,...,0.008271,0.013131,0.039529,0.199201,0.012169,0.043606,0.029219,0.032052,0.022264,0.158456
1991-01-01,0.050069,0.169180,0.092059,0.010839,0.922185,0.014070,0.177852,0.100253,0.013249,0.427301,...,0.008230,0.013551,0.036796,0.189353,0.012997,0.043209,0.028915,0.031793,0.019700,0.156578
1992-01-01,0.048945,0.162535,0.089719,0.011846,0.899414,0.013174,0.181219,0.103134,0.013791,0.416479,...,0.008058,0.013557,0.036438,0.156331,0.013389,0.040847,0.027608,0.030698,0.018933,0.153208
1993-01-01,0.049420,0.165972,0.089637,0.012462,0.868767,0.012748,0.184745,0.106892,0.013975,0.403464,...,0.008281,0.013570,0.036788,0.138956,0.014523,0.040243,0.026313,0.029838,0.018437,0.154310
1994-01-01,0.049292,0.168882,0.090037,0.012722,0.855950,0.012615,0.189415,0.110500,0.014431,0.392040,...,0.008498,0.013704,0.037235,0.116784,0.015510,0.039922,0.026310,0.029047,0.018239,0.155681


We interpret it as : in 1990, we should rationate the US price level by a scale of 0.9012 to compare it to the EU taking the Penn effect into account.

In [12]:
#Building the theoretical exchange rate

df_PPP

CURRENCY,AUD,BRL,CAD,CHF,CLP,CZK,EUR,GBP,HUF,IDR,...,INR,JPY,MXN,NOK,NZD,PLN,RUB,SEK,SGD,ZAR
TIME_PERIOD,,,,,,,,,,,,,,,,,,,,,
1995-01-01,1.310302,0.539108,1.209018,1.922675,278.809422,11.152963,0.875512,0.712913,61.472314,510.861176,...,7.952394,174.247214,3.006706,9.069835,1.466538,1.163351,1.715170,9.348993,1.006373,2.034084
1996-01-01,1.313356,0.627126,1.207172,1.887199,278.469518,12.071769,0.882957,0.712844,73.855397,545.204682,...,8.400987,170.351722,3.740912,8.973256,1.472926,1.350763,2.455975,9.245636,1.003165,2.157757
1997-01-01,1.307709,0.664144,1.200327,1.844032,282.866287,12.902515,0.867835,0.708637,87.377995,603.340128,...,8.793420,168.308686,4.329651,9.055966,1.451281,1.512912,2.778421,9.329317,0.996859,2.293788
1998-01-01,1.300774,0.689104,1.184866,1.827602,281.901592,13.986306,0.862638,0.720671,97.816595,1045.728708,...,9.392232,166.393737,4.954420,9.311803,1.449162,1.654902,3.258399,9.395336,0.972681,2.451477
1999-01-01,1.296950,0.733919,1.190810,1.802836,284.049363,14.229350,0.865406,0.725819,104.545883,1177.158453,...,9.545346,162.035740,5.633670,9.283739,1.434660,1.737656,5.539641,9.305900,0.924793,2.583792
2000-01-01,1.312261,0.757895,1.228197,1.788660,290.423477,14.327536,0.860905,0.704372,110.049076,1386.451585,...,9.674133,154.687365,6.078239,9.083460,1.444868,1.832541,7.301404,9.161129,0.939247,2.758554
2001-01-01,1.328086,0.802171,1.220053,1.768621,296.176351,14.323590,0.863926,0.694551,114.418570,1549.758991,...,9.765331,149.639771,6.329170,9.175302,1.473448,1.837258,8.322817,9.402445,0.901897,2.908249
2002-01-01,1.336490,0.867291,1.229333,1.709760,301.387054,14.458321,0.866973,0.689877,118.141282,1616.020707,...,9.973197,143.774205,6.553697,9.056730,1.468962,1.799795,9.273504,9.413655,0.880130,3.222802
2003-01-01,1.352188,0.970343,1.225777,1.715154,308.261553,14.177038,0.887261,0.695928,122.020996,1671.694194,...,10.158382,139.581668,6.647339,9.198149,1.495284,1.804595,9.866787,9.486667,0.847563,3.362538


The signal we want to build is the deviation to the theoretical exchange rate we've computed $signal = MovAvg(St_{market}, ndays) / St_{theoretical}$.  
We assume mean reversion and go long bottom 7 and short top 7 (3 for G10)

In [7]:
from data import spt_dol_import

mapping = {'AUSTDOL(ER)': "AUD", 'BRACRUZ(ER)':'BRL', 'CNDOLLR(ER)':'CAD', 'CHILPES(ER)':'CLP',
       'USEURSP(ER)':"EUR", 'HUNFORT(ER)':"HUF", 'INDRUPE(ER)':"INR", 'INDORUP(ER)':"IDR",
       'ISRSHEK(ER)':"ILS", 'JAPAYEN(ER)':"JPY", 'MEXPESO(ER)':"MXN", 'NZDOLLR(ER)':"NZD",
       'NORKRON(ER)':"NOK", 'POLZLOT(ER)':"PLN", 'CISRUBM(ER)':"RUB", 'SINGDOL(ER)':"SGD",
       'COMRAND(ER)':"ZAR", 'SWEKRON(ER)':"SEK", 'SWISSFR(ER)':"CHF", 'CZECHCM(ER)':"CZK", 'GBP(ER)':"GBP"}

spt_dol = spt_dol_import()
spt_dol_ma = spt_dol.rolling(30).mean().rename(mapping, axis=1) #The 30 days choice is tipically arbitrary, we might try other frames
spt_dol_ma[28:32]

,AUD,BRL,CAD,CLP,EUR,HUF,INR,IDR,ILS,JPY,...,NZD,NOK,PLN,RUB,SGD,ZAR,SEK,CHF,CZK,GBP
Code,,,,,,,,,,,,,,,,,,,,,
1998-02-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1998-02-11,0.662274,0.892272,0.694452,0.002211,1.080236,0.004846,0.025570,0.000109,0.279415,0.007807,...,0.581636,0.133276,0.283185,0.165933,0.580898,0.202703,0.124552,0.680041,0.028446,1.637320
1998-02-12,0.663158,0.892002,0.694341,0.002210,1.079947,0.004844,0.025579,0.000108,0.279256,0.007820,...,0.581786,0.133184,0.283138,0.165884,0.581306,0.202613,0.124480,0.680144,0.028453,1.637138
1998-02-13,0.663882,0.891701,0.694069,0.002209,1.079471,0.004841,0.025590,0.000106,0.279108,0.007834,...,0.581974,0.133063,0.283052,0.165842,0.581704,0.202534,0.124405,0.680186,0.028456,1.636983


In [8]:
#The median rebalancement try to avoid any noise from the beginning of a period
rebal_mth = spt_dol.index.to_series().groupby(spt_dol.index.to_period("M")).min() #Monthly rebalancement
rebal_qtr = spt_dol.index.to_series().groupby(spt_dol.index.to_period('Q')).min() #Quarterly rebalancement
rebal_qmed = spt_dol.index.to_series().groupby(spt_dol.index.to_period("Q")).median().dt.normalize() #Quarterly median date

In [ ]:
#We reindex the data to get a dataframe with observations at the right intervals
spt_mth = spt_dol.reindex(rebal_mth, method="ffill")
spt_qtr = spt_dol.reindex(rebal_qtr, method ="ffill")
spt_qmed = spt_dol.reindex(rebal_qmed, method="ffill")

df_PPPm = df_PPP.reindex(rebal_mth, method='ffill')
df_PPPq = df_PPP.reindex(rebal_qtr, method="ffill")
df_PPPqmed = df_PPP.reindex(rebal_qmed, method="ffill")

df_GDPm = df_GDP.reindex(rebal_mth, method="ffill")
df_GDPq = df_GDP.reindex(rebal_qtr, method="ffill")
df_GDPqmed = df_GDP.reindex(rebal_qmed, method="ffill")

We might aim to find a long term precise value by running a regression in sample of the form $St = ß0 + ßprices * P(USD)/P(f) + ßgdp * GDP(f)/GDP(usd)$ or $st = ß0 + ßprices * p(USD)-p(f) + ßgdp * gdp(f)-gdp(usd)$